### DISCUSSION TITLE
Oqtane 3.1 IIS Deployment

### DISCUSSION THREAD

**Discussion by Sharafudeen** [oqtane/oqtane.framework#2150] [Q&A]
Hi,

I've deployed Oqtane 3.1 to IIS but it is not working at all. Why it is creating appsettings.production.json & web.Release.config? Which configurations Oqtane can use it at runtime?

Also, whenever run oqtane with different urls (https://localhost:44357 or https://localhost:5000) Oqtane.Server.exe file then it creates Alias in the database.

The log shows, there is an issues in the following path,

1. \oqtane.framework-3.1.0\Oqtane.Server\Infrastructure\Middleware\JwtMiddleware.cs (line: 65)
2. \oqtane.framework-3.1.0\Oqtane.Server\Infrastructure\Middleware\TenantMiddleware.cs (line: 58)

**Comment by sbwalker** [comment]
Did you follow all of the instructions exactly as described here: 

https://www.oqtane.org/blog/!/24/installing-oqtane-on-iis

Most issues related to installing on IIS are due to misconfiguration.

**Reply by Sharafudeen** [reply]
Yes I've followed the same as described in the link. Actually I've created my Oqtane application under my XYZ.com/mysite and created new application pool for my Oqtane application and configured as per the link you provided.

**Reply by sbwalker** [reply]
I am not sure I understand the reference to xyz.com/mysite. Is this a local IIS development environment or a public IIS server? Did you create a new IIS site for Oqtane or are you trying to create a virtual directory under an existing site? How did you configure the host name - local hosts file or actual DNS?

**Reply by Sharafudeen** [reply]
xyz.com/mysite this is public DNS but it can accessible only intranet (within the organization). I've created mysite as virtual directory under an existing site XYZ.com. In my case, organization won't allow me to create root site. I didn't configure anything in "local hosts file".

**Reply by Sharafudeen** [reply]
**Note:**  I've multiple sites under xyz.com. Many of them ASPNET Core applications and their application pool configured in dotnet v4. But only for Oqtane application and it's application pool is "No Managed Code" configured.

**Reply by sbwalker** [reply]
It sounds like this is shared Intranet IIS server which is running a variety of applications. And it sounds like you do not have administrative rights to the server (ie. you cannot create new root level sites ). So I am wondering how you were able to perform steps 1-3 in the installation instructions - as these require administrative access? If any of the options described in steps 1-3 are not configured correctly, you will have difficulty running Oqtane. In addition since this is not a local IIS server you will also need to verify the loop back requirements - as the client needs to be able to makes calls to the API over HTTP. There are other threads in which discuss this in more detail.

**Reply by sbwalker** [reply]
This is another good list of installation instructions which may help with your scenario:

https://azing.org/oqtane/r/7mk6DNd0

**Reply by Sharafudeen** [reply]
I've full rights to create website under IIS. But for our company policy we should not create a root site since it is already created. I'm not trying to newly install Oqtane in my xyz.com/mysite but actually I've published from VS2022 and copied files to mysite. The database in my development environment for Oqtane is same as mysite. There's no special in the deployment!

**Reply by sbwalker** [reply]
The installation instructions Step 5 indicate to use an official Install package of Oqtane - NOT to publish from Visual Studio. This is because the official package has some customizations.

**Reply by Sharafudeen** [reply]
Just wanted to know.. Oqtane should be flexible with deployment right? This is because, each organization or company has different types of infrastructure and their development and deployment might be different.. In my case, we have DMZ servers and our deployment admins doing manual. What I'm expecting is to move published Oqtane framework sources with modules DLLs from development machine to quality server.

**Reply by sbwalker** [reply]
Yes, Oqtane is flexible with deployment - however the more customized your installation, the more difficult it is for the community to offer assistance or advice when you run into issues. If you want to rebuild the Oqtane source you should use the release.cmd file in the Oqtane.Package folder as that is the build script used for official releases. However, as I mentioned previously you don't need to rebuild the source yourself if you use the Install package. And you can deploy your custom modules by copying the nuget packages into the appropriate folder location ( this blog describes deploying extensions https://www.oqtane.org/blog/!/38/site-migrations---deployment )

**Reply by Sharafudeen** [reply]
If I use release.cmd or installation package for existing database then it will affects my previously created tables and it's data. How do I prevent database migration while installing or running the release.cmd?

**Reply by Sharafudeen** [reply]
I've downloaded latest release **Oqtane.Framework.3.1.0.Install.zip** file and extract to the IIS and tried to install but it log as follows,

[2022-04-25 05:19:50+00:00] [Error] Connection ID "17870283327311708858", Request ID "800002bb-0001-f800-b63f-84710c7967bb": An unhandled exception was thrown by the application.    at Microsoft.AspNetCore.Authentication.RemoteAuthenticationHandler`1.HandleRequestAsync()
   at Microsoft.AspNetCore.Authentication.AuthenticationMiddleware.Invoke(HttpContext context)
   at Oqtane.Infrastructure.JwtMiddleware.Invoke(HttpContext context) in C:\Users\Shaun.Walker\source\repos\sbwalker\oqtane.framework\Oqtane.Server\Infrastructure\Middleware\JwtMiddleware.cs:line 65
   at Oqtane.Infrastructure.TenantMiddleware.Invoke(HttpContext context) in C:\Users\Shaun.Walker\source\repos\sbwalker\oqtane.framework\Oqtane.Server\Infrastructure\Middleware\TenantMiddleware.cs:line 58
   at Microsoft.AspNetCore.Localization.RequestLocalizationMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Builder.Extensions.UsePathBaseMiddleware.InvokeCore(HttpContext context, String matchedPath, String remainingPath)
   at Microsoft.AspNetCore.Server.IIS.Core.IISHttpContextOfT`1.ProcessRequestAsync()

**Reply by sbwalker** [reply]
Is this a new install or upgrade? I am asking because previously it sounded like it was a new install - but now you mentioned that you have an existing database - which would indicate that you installed the framework previously. It's really difficult to help without knowing all of the specific details of the environment and the process which was followed. The error message is not helpful - it seems to point to an issue in the tenant middleware which I have never seen. I can only guess but it might be because your database has invalid Alias entries in it... or it might be related to the virtual directory setup (I have never tried that before)... or it might be something else entirely.

**Reply by Sharafudeen** [reply]
Actually I've downloaded from Oqtane framework GitHub releases link with mentioned version. In my previous database I've deleted all tables and no tables exists. Then, I've browse XYZ.com/mysite and it is not working. When I checked the log under Content folder and it created error file with above error which I've shared.

**Note:** I didn't change anything in the Host file in the windows server and no connectionstring mentioned in Oqtane appsettings.json

**Reply by sbwalker** [reply]
Ok... so you are saying that you are now using the Oqtane.Framework.3.1.0.Install.zip package. This means you are using the standard [appsettings.json](https://github.com/oqtane/oqtane.framework/blob/dev/Oqtane.Server/appsettings.json) file with no customizations. So if that is true then there is no ConnectionString value specified. So when the application starts up, it will recognize that the database has not yet been configured and it should display the Installation Wizard. Are you saying the Installation Wizard is not being displayed?

**Reply by Sharafudeen** [reply]
Yes.. Installation Wizard is not displayed.. instead that error logged.

**Reply by sbwalker** [reply]
You are certain that your ConnectionString is blank in appsettings.json?

**Reply by Sharafudeen** [reply]
Yes ConnectionString is blank.

**Reply by sbwalker** [reply]
I have a feeling that the problems you are experiencing are not related to Oqtane - but rather related to how Blazor behaves when installed on a virtual directory. I have never tried to run Blazor in a virtual directory so I would encourage you to explore the documentation from Microsoft on this topic: https://docs.microsoft.com/en-us/aspnet/core/blazor/host-and-deploy/?view=aspnetcore-6.0&tabs=visual-studio It appears that they do not recommend using virtual directories - but if you have no choice then there are a variety of tweaks you may need to make to web.config, app base path, etc...

**Reply by Sharafudeen** [reply]
Thanks for the clarification! Could you please tell me what are the places needs to be modified in Oqtane framework?

**Reply by sbwalker** [reply]
I have no idea... and unfortunately I do not have time to troubleshoot this configuration. However, if you could please share your own findings it would be most appreciated.

**Reply by sbwalker** [reply]
The first section in the Microsoft link I shared says "avoid the use of IIS's virtual directories for hosting multiple apps" - so are you trying to use a virtual directory or a sub-application? If you are trying to use a sub-application it is still not going to work because Oqtane is not using tilde (~) notation for its static resources, and the multi-tenant resolution logic does not include support for a sub-app path. Your best approach at this point is to create a dedicated site for installation.

**Reply by Sharafudeen** [reply]
Actually we have xyz.com and I'm creating an application mysite under xyz.com/mysite. Mysite physical path is same as xyz.com application but mysite has different configuration such as Application pool and so on.

I really worries.. how Blazor and Oqtane didn't meet such requirements? Basically this feature should be there.. I'm planning to commit Oqtane in our organization but now thinking about alternative..

**Reply by sbwalker** [reply]
@Sharafudeen I am very confused by the following phrase in your last message.... "Mysite physical path is same as xyz.com" - are you saying that xyz.com and xyz.com/mysite point to the same physical location on disk? Does this mean that both xyz.com and xyz.com/mysite are running the same code? And is that code Oqtane?

**Reply by Sharafudeen** [reply]
Sorry.. No! I've created a folder "mysite" inside xyz.com site and it is configured as No Managed Code in the application pool but xyz.com site has different application pool and configuration.

**Reply by sbwalker** [reply]
@Sharafudeen here is a SO post of someone else having an issue with running Blazor in a sub application - https://stackoverflow.com/questions/62026817/deploy-blazor-app-to-a-iis-sub-application. It does appear to be possible but there would be a LOT of changes and testing required to support it in Oqtane, as it affects all static resources, tenant resolution, cookie authentication , etc.... 

I updated the IIS installation instructions to make some aspects more clear: https://www.oqtane.org/blog/!/24/installing-oqtane-on-iis

**Reply by Sharafudeen** [reply]
@sbwalker I can find the required places to modify but it will be customized Oqtane solution. I don't want to do that! I would like to keep Oqtane framework as an official release. This is because long-term maintenance! I would recommend, please make it Oqtane framework is more flexible such kind of use cases. Many of an organization might not allowed to create domain and open the port for "sub application" (In our case, we have more than 50 sub applications) and if they do it then they might have to invest for other setups such as security certificate, server maintenance, etc., There are indirect efforts and costs involved.. Hope you will understand.. Thanks!
